File: scpn-fusion-rs/crates/fusion-physics/src/hall_mhd_pulsed.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SCPN Fusion Core — Pulsed Hall-MHD
//! Axisymmetric Ono Eq. 8 Hall-MHD flux carrier.

pub const MU_0: f64 = 4.0e-7 * std::f64::consts::PI;

#[derive(Debug, Clone, PartialEq)]
pub struct HallMhdPulsedConfig {
    pub rho_m: Vec<f64>,
    pub tau_psi_s: f64,
    pub r_null_m: f64,
    pub electron_temperature_ev: f64,
    pub z_eff: f64,
    pub ln_lambda: f64,
    pub hall_scale: f64,
}

#[derive(Debug, Clone, PartialEq)]
pub struct HallMhdPulsedState {
    pub t_s: f64,
    pub psi: Vec<f64>,
    pub b_z: Vec<f64>,
    pub e_theta: Vec<f64>,
    pub j_theta: Vec<f64>,
    pub energy_proxy_j_m: f64,
    pub hall_drive_l2: f64,
    pub resistive_sink_l2: f64,
    pub damping_sink_l2: f64,
    pub source_residual_linf: f64,
    pub external_parity_status: &'static str,
}

fn require_positive(name: &str, value: f64) -> Result<f64, String> {
    if !value.is_finite() {
        return Err(format!("{name} must be finite"));
    }
    if value <= 0.0 {
        return Err(format!("{name} must be positive"));
    }
    Ok(value)
}

fn validate_grid(rho: &[f64]) -> Result<(), String> {
    if rho.len() < 3 {
        return Err("rho_m must contain at least three points".to_string());
    }
    if rho[0] < 0.0 || !rho[0].is_finite() {
        return Err("rho_m must start at a non-negative finite radius".to_string());
    }
    for pair in rho.windows(2) {
        if !pair[1].is_finite() || pair[1] <= pair[0] {
            return Err("rho_m must be strictly increasing".to_string());
        }
    }
    Ok(())
}

fn validate_profile(name: &str, values: &[f64], expected: usize) -> Result<(), String> {
    if values.len() != expected {
        return Err(format!("{name} must have length {expected}"));
    }
    if values.iter().any(|value| !value.is_finite()) {
        return Err(format!("{name} must be finite"));
    }
    Ok(())
}

pub fn spitzer_resistivity_ohm_m(
    temperature_ev: f64,
    z_eff: f64,
    ln_lambda: f64,
) -> Result<f64, String> {
    let temperature = require_positive("temperature_eV", temperature_ev)?;
    let z = require_positive("Z_eff", z_eff)?;
    let log = require_positive("ln_lambda", ln_lambda)?;
    Ok(1.65e-9 * z * log / temperature.powf(1.5))
}

pub fn faraday_e_theta_from_b_ramp(rho_m: &[f64], d_b_ext_dt_t_s: f64) -> Result<Vec<f64>, String> {
    validate_grid(rho_m)?;
    if !d_b_ext_dt_t_s.is_finite() {
        return Err("dB_ext_dt must be finite".to_string());
    }
    Ok(rho_m.iter().map(|r| -0.5 * r * d_b_ext_dt_t_s).collect())
}

pub fn axial_field_from_flux(rho_m: &[f64], psi: &[f64]) -> Result<Vec<f64>, String> {
    validate_grid(rho_m)?;
    validate_profile("psi", psi, rho_m.len())?;
    let gradient = gradient_edge_order2(rho_m, psi);
    let mut field = vec![0.0; psi.len()];
    for index in 1..psi.len() {
        field[index] = gradient[index] / rho_m[index];
    }
    field[0] = field[1];
    Ok(field)
}

pub fn initial_hall_mhd_pulsed_state(
    cfg: &HallMhdPulsedConfig,
    psi: Vec<f64>,
    e_theta: Vec<f64>,
    j_theta: Vec<f64>,
) -> Result<HallMhdPulsedState, String> {
    validate_config(cfg)?;
    validate_profile("psi", &psi, cfg.rho_m.len())?;
    validate_profile("E_theta", &e_theta, cfg.rho_m.len())?;
    validate_profile("J_theta", &j_theta, cfg.rho_m.len())?;
    let b_z = axial_field_from_flux(&cfg.rho_m, &psi)?;
    Ok(HallMhdPulsedState {
        t_s: 0.0,
        energy_proxy_j_m: magnetic_energy_proxy(&cfg.rho_m, &psi),
        hall_drive_l2: l2(&cfg.rho_m, &scale(&e_theta, cfg.hall_scale * cfg.r_null_m)),
        resistive_sink_l2: l2(&cfg.rho_m, &scale(&j_theta, eta(cfg)?)),
        damping_sink_l2: l2(&cfg.rho_m, &scale(&psi, 1.0 / cfg.tau_psi_s)),
        source_residual_linf: 0.0,
        external_parity_status: "blocked_missing_public_same_case_reference",
        psi,
        b_z,
        e_theta,
        j_theta,
    })
}

pub fn step_hall_mhd_pulsed(
    state: &HallMhdPulsedState,
    cfg: &HallMhdPulsedConfig,
    e_theta: Vec<f64>,
    j_theta: Vec<f64>,
    dt_s: f64,
) -> Result<HallMhdPulsedState, String> {
    validate_config(cfg)?;
    validate_profile("state.psi", &state.psi, cfg.rho_m.len())?;
    validate_profile("E_theta", &e_theta, cfg.rho_m.len())?;
    validate_profile("J_theta", &j_theta, cfg.rho_m.len())?;
    let dt = require_positive("dt_s", dt_s)?;
    let eta_value = eta(cfg)?;
    let source = ono_source(cfg, &e_theta, &j_theta, eta_value);
    let denominator = 1.0 + dt / cfg.tau_psi_s;
    let psi = state
        .psi
        .iter()
        .zip(source.iter())
        .map(|(old, src)| (old + dt * src) / denominator)
        .collect::<Vec<_>>();
    let residual = psi
        .iter()
        .zip(state.psi.iter())
        .zip(source.iter())
        .map(|((new_value, old_value), src)| {
            (new_value - old_value) / dt - src + new_value / cfg.tau_psi_s
        })
        .collect::<Vec<_>>();
    let scale_value = source
        .iter()
        .chain(psi.iter())
        .map(|value| value.abs())
        .fold(1.0_f64, f64::max);
    let b_z = axial_field_from_flux(&cfg.rho_m, &psi)?;
    Ok(HallMhdPulsedState {
        t_s: state.t_s + dt,
        energy_proxy_j_m: magnetic_energy_proxy(&cfg.rho_m, &psi),
        hall_drive_l2: l2(&cfg.rho_m, &scale(&e_theta, cfg.hall_scale * cfg.r_null_m)),
        resistive_sink_l2: l2(&cfg.rho_m, &scale(&j_theta, eta_value)),
        damping_sink_l2: l2(&cfg.rho_m, &scale(&psi, 1.0 / cfg.tau_psi_s)),
        source_residual_linf: residual
            .iter()
            .map(|value| value.abs())
            .fold(0.0_f64, f64::max)
            / scale_value,
        external_parity_status: "blocked_missing_public_same_case_reference",
        psi,
        b_z,
        e_theta,
        j_theta,
    })
}

fn validate_config(cfg: &HallMhdPulsedConfig) -> Result<(), String> {
    validate_grid(&cfg.rho_m)?;
    require_positive("tau_psi_s", cfg.tau_psi_s)?;
    require_positive("r_null_m", cfg.r_null_m)?;
    require_positive("electron_temperature_eV", cfg.electron_temperature_ev)?;
    require_positive("Z_eff", cfg.z_eff)?;
    require_positive("ln_lambda", cfg.ln_lambda)?;
    if !cfg.hall_scale.is_finite() {
        return Err("hall_scale must be finite".to_string());
    }
    Ok(())
}

fn eta(cfg: &HallMhdPulsedConfig) -> Result<f64, String> {
    spitzer_resistivity_ohm_m(cfg.electron_temperature_ev, cfg.z_eff, cfg.ln_lambda)
}

fn ono_source(
    cfg: &HallMhdPulsedConfig,
    e_theta: &[f64],
    j_theta: &[f64],
    eta_value: f64,
) -> Vec<f64> {
    e_theta
        .iter()
        .zip(j_theta.iter())
        .map(|(e, j)| cfg.hall_scale * cfg.r_null_m * e - eta_value * j)
        .collect()
}

fn gradient_edge_order2(x: &[f64], y: &[f64]) -> Vec<f64> {
    let n = x.len();
    let mut gradient = vec![0.0; n];
    gradient[0] = (-3.0 * y[0] + 4.0 * y[1] - y[2]) / (x[2] - x[0]);
    gradient[n - 1] = (3.0 * y[n - 1] - 4.0 * y[n - 2] + y[n - 3]) / (x[n - 1] - x[n - 3]);
    for index in 1..(n - 1) {
        gradient[index] = (y[index + 1] - y[index - 1]) / (x[index + 1] - x[index - 1]);
    }
    gradient
}

fn trapezoid(x: &[f64], y: &[f64]) -> f64 {
    x.windows(2)
        .zip(y.windows(2))
        .map(|(xw, yw)| 0.5 * (yw[0] + yw[1]) * (xw[1] - xw[0]))
        .sum()
}

fn magnetic_energy_proxy(rho: &[f64], psi: &[f64]) -> f64 {
    let integrand = rho
        .iter()
        .zip(psi.iter())
        .map(|(r, p)| 0.5 * p * p * 2.0 * std::f64::consts::PI * r / MU_0)
        .collect::<Vec<_>>();
    trapezoid(rho, &integrand)
}

fn l2(rho: &[f64], values: &[f64]) -> f64 {
    let integrand = rho
        .iter()
        .zip(values.iter())
        .map(|(r, value)| value * value * 2.0 * std::f64::consts::PI * r)
        .collect::<Vec<_>>();
    trapezoid(rho, &integrand).max(0.0).sqrt()
}

fn scale(values: &[f64], factor: f64) -> Vec<f64> {
    values.iter().map(|value| value * factor).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn cfg() -> HallMhdPulsedConfig {
        HallMhdPulsedConfig {
            rho_m: (0..129).map(|i| i as f64 * 0.4 / 128.0).collect(),
            tau_psi_s: 2.0e-6,
            r_null_m: 0.2,
            electron_temperature_ev: 5_000.0,
            z_eff: 1.0,
            ln_lambda: 17.0,
            hall_scale: 1.0,
        }
    }

    #[test]
    fn faraday_drive_matches_circular_loop_formula() {
        let field = faraday_e_theta_from_b_ramp(&[0.0, 0.1, 0.2], 2.0e4).unwrap();
        assert!((field[1] + 1000.0).abs() < 1.0e-10);
        assert!((field[2] + 2000.0).abs() < 1.0e-10);
    }

    #[test]
    fn spitzer_temperature_scaling() {
        let eta_100 = spitzer_resistivity_ohm_m(100.0, 1.0, 17.0).unwrap();
        let eta_400 = spitzer_resistivity_ohm_m(400.0, 1.0, 17.0).unwrap();
        assert!((eta_400 - eta_100 / 8.0).abs() / eta_400 < 1.0e-14);
    }

    #[test]
    fn implicit_damping_matches_closed_form_decay() {
        let mut case = cfg();
        case.hall_scale = 0.0;
        let psi = case.rho_m.iter().map(|r| 1.0 + r).collect::<Vec<_>>();
        let zeros = vec![0.0; case.rho_m.len()];
        let state = initial_hall_mhd_pulsed_state(&case, psi.clone(), zeros.clone(), zeros.clone())
            .unwrap();
        let next = step_hall_mhd_pulsed(&state, &case, zeros.clone(), zeros, 1.0e-7).unwrap();
        let expected_factor = 1.0 / (1.0 + 1.0e-7 / case.tau_psi_s);
        assert!((next.psi[7] - psi[7] * expected_factor).abs() < 1.0e-12);
    }

    #[test]
    fn hall_drive_increases_flux() {
        let mut case = cfg();
        case.tau_psi_s = 1.0e9;
        let psi = case.rho_m.iter().map(|r| 1.0 + r).collect::<Vec<_>>();
        let zeros = vec![0.0; case.rho_m.len()];
        let drive = vec![2.5; case.rho_m.len()];
        let state =
            initial_hall_mhd_pulsed_state(&case, psi, drive.clone(), zeros.clone()).unwrap();
        let next = step_hall_mhd_pulsed(&state, &case, drive, zeros, 1.0e-8).unwrap();
        assert!(next
            .psi
            .iter()
            .zip(state.psi.iter())
            .all(|(new_value, old_value)| new_value > old_value));
        assert!(next.hall_drive_l2 > 0.0);
    }

    #[test]
    fn small_hall_limit_recovers_resistive_sink() {
        let mut case = cfg();
        case.hall_scale = 0.0;
        case.tau_psi_s = 1.0e9;
        let psi = vec![1.0; case.rho_m.len()];
        let drive = vec![10.0; case.rho_m.len()];
        let current = vec![2.0e5; case.rho_m.len()];
        let state =
            initial_hall_mhd_pulsed_state(&case, psi, drive.clone(), current.clone()).unwrap();
        let next = step_hall_mhd_pulsed(&state, &case, drive, current, 1.0e-8).unwrap();
        let expected =
            (1.0 - 1.0e-8 * eta(&case).unwrap() * 2.0e5) / (1.0 + 1.0e-8 / case.tau_psi_s);
        assert!((next.psi[9] - expected).abs() < 1.0e-14);
        assert_eq!(next.hall_drive_l2, 0.0);
    }
}

File: src/scpn_fusion/core/hall_mhd_pulsed.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Fusion Core — MIF/FRC Pulsed Hall-MHD
"""Axisymmetric pulsed Hall-MHD flux carrier for the MIF/FRC lane.

The accepted contract evolves the flattened Ono Eq. 8 carrier

``dpsi/dt = -psi/tau_psi + R_null E_theta - eta_spitzer J_theta``

on the accepted FRC radial grid. It is a production flux-evolution surface for
MIF trigger/replay work, not a claim of full 2D two-fluid Hall-MHD or Gkeyll
same-case parity.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from typing import Literal, TypeAlias

import numpy as np
from numpy.typing import NDArray
from scipy.integrate import trapezoid

from .frc_rigid_rotor import FRCEquilibriumState, MU_0

FloatArray: TypeAlias = NDArray[np.float64]
ScalarCallable: TypeAlias = Callable[[float], float]
ProfileCallable: TypeAlias = Callable[[float, FloatArray], FloatArray | float]
IntegratorName: TypeAlias = Literal["implicit_be", "imex_rk2"]
ClosureName: TypeAlias = Literal["axisymmetric_ono_flux", "adiabatic_electron"]


@dataclass(frozen=True)
class HallMHDPulsedConfig:
    """Configuration for the accepted axisymmetric pulsed Hall-MHD carrier."""

    equilibrium: FRCEquilibriumState
    B_ext_t: ScalarCallable
    tau_psi_s: float
    electron_temperature_eV: float
    E_theta_t: ProfileCallable | None = None
    J_theta_t: ProfileCallable | None = None
    Z_eff: float = 1.0
    ln_lambda: float = 17.0
    hall_scale: float = 1.0
    closure: ClosureName = "axisymmetric_ono_flux"
    time_integrator: IntegratorName = "implicit_be"


@dataclass(frozen=True)
class HallMHDPulsedState:
    """State of the axisymmetric pulsed Hall-MHD flux carrier."""

    t_s: float
    psi: FloatArray
    B_z: FloatArray
    E_theta: FloatArray
    J_theta: FloatArray
    R_null_m: float
    energy_proxy_J_m: float
    hall_drive_l2: float
    resistive_sink_l2: float
    damping_sink_l2: float
    source_residual_linf: float
    closure_status: str
    external_parity_status: str


def spitzer_resistivity_ohm_m(
    temperature_eV: FloatArray | float,
    *,
    Z_eff: float = 1.0,
    ln_lambda: float = 17.0,
) -> FloatArray:
    """Return NRL-style Spitzer resistivity ``1.65e-9 Z lnLambda / T_e^1.5``."""
    temperature = np.asarray(temperature_eV, dtype=np.float64)
    if np.any(~np.isfinite(temperature)) or np.any(temperature <= 0.0):
        raise ValueError("temperature_eV must be positive and finite")
    z_eff = _require_positive("Z_eff", Z_eff)
    coulomb_log = _require_positive("ln_lambda", ln_lambda)
    return np.asarray(1.65e-9 * z_eff * coulomb_log / np.power(temperature, 1.5), dtype=np.float64)


def faraday_e_theta_from_b_ramp(
    rho_m: FloatArray,
    B_ext_t: ScalarCallable,
    t_s: float,
    *,
    derivative_dt_s: float = 1.0e-9,
) -> FloatArray:
    """Return circular-loop Faraday drive ``E_theta = -r/2 dB_ext/dt``."""
    rho = _validate_grid(rho_m)
    dt = _require_positive("derivative_dt_s", derivative_dt_s)
    time = _require_non_negative("t_s", t_s)
    if time >= dt:
        dB_dt = (B_ext_t(time + dt) - B_ext_t(time - dt)) / (2.0 * dt)
    else:
        dB_dt = (B_ext_t(time + dt) - B_ext_t(time)) / dt
    dB_dt = _require_finite("dB_ext_dt", dB_dt)
    return np.asarray(-0.5 * rho * dB_dt, dtype=np.float64)


def axial_field_from_flux(rho_m: FloatArray, psi: FloatArray) -> FloatArray:
    """Return axisymmetric ``B_z = (1/r) dpsi/dr`` with finite-axis handling."""
    rho = _validate_grid(rho_m)
    flux = _validate_profile("psi", psi, rho.size)
    dpsi_dr = np.gradient(flux, rho, edge_order=2)
    field = np.empty_like(flux)
    field[1:] = dpsi_dr[1:] / rho[1:]
    field[0] = field[1]
    return np.asarray(field, dtype=np.float64)


def initial_hall_mhd_pulsed_state(config: HallMHDPulsedConfig) -> HallMHDPulsedState:
    """Construct the initial Hall-MHD carrier state from an FRC equilibrium."""
    cfg = _validate_config(config)
    rho = cfg.equilibrium.rho
    psi = np.asarray(cfg.equilibrium.psi, dtype=np.float64).copy()
    e_theta = _resolve_e_theta(cfg, 0.0)
    j_theta = _resolve_j_theta(cfg, 0.0)
    return HallMHDPulsedState(
        t_s=0.0,
        psi=psi,
        B_z=axial_field_from_flux(rho, psi),
        E_theta=e_theta,
        J_theta=j_theta,
        R_null_m=float(cfg.equilibrium.R_null),
        energy_proxy_J_m=_magnetic_energy_proxy(rho, psi),
        hall_drive_l2=_l2(rho, cfg.hall_scale * cfg.equilibrium.R_null * e_theta),
        resistive_sink_l2=_l2(rho, _eta_profile(cfg, rho) * j_theta),
        damping_sink_l2=_l2(rho, psi / cfg.tau_psi_s),
        source_residual_linf=0.0,
        closure_status=cfg.closure,
        external_parity_status=gkeyll_small_hall_acceptance_status()["status"],
    )


def step_hall_mhd_pulsed(
    state: HallMHDPulsedState,
    config: HallMHDPulsedConfig,
    dt_s: float,
) -> HallMHDPulsedState:
    """Advance one accepted axisymmetric Ono Eq. 8 carrier step."""
    cfg = _validate_config(config)
    dt = _require_positive("dt_s", dt_s)
    _validate_state(state, cfg)
    if cfg.time_integrator == "implicit_be":
        next_psi, source = _implicit_be_step(state.psi, state.t_s, cfg, dt)
    elif cfg.time_integrator == "imex_rk2":
        next_psi, source = _imex_rk2_step(state.psi, state.t_s, cfg, dt)
    else:
        raise ValueError("time_integrator must be 'implicit_be' or 'imex_rk2'")
    rho = cfg.equilibrium.rho
    e_theta = _resolve_e_theta(cfg, state.t_s + dt)
    j_theta = _resolve_j_theta(cfg, state.t_s + dt)
    residual = (next_psi - state.psi) / dt - source + next_psi / cfg.tau_psi_s
    scale = max(float(np.max(np.abs(source))), float(np.max(np.abs(next_psi / cfg.tau_psi_s))), 1.0)
    return HallMHDPulsedState(
        t_s=state.t_s + dt,
        psi=next_psi,
        B_z=axial_field_from_flux(rho, next_psi),
        E_theta=e_theta,
        J_theta=j_theta,
        R_null_m=float(cfg.equilibrium.R_null),
        energy_proxy_J_m=_magnetic_energy_proxy(rho, next_psi),
        hall_drive_l2=_l2(rho, cfg.hall_scale * cfg.equilibrium.R_null * e_theta),
        resistive_sink_l2=_l2(rho, _eta_profile(cfg, rho) * j_theta),
        damping_sink_l2=_l2(rho, next_psi / cfg.tau_psi_s),
        source_residual_linf=float(np.max(np.abs(residual)) / scale),
        closure_status=cfg.closure,
        external_parity_status=gkeyll_small_hall_acceptance_status()["status"],
    )


def run_hall_mhd_pulsed(
    initial: HallMHDPulsedState,
    config: HallMHDPulsedConfig,
    dt_s: float,
    n_steps: int,
) -> tuple[HallMHDPulsedState, ...]:
    """Run a supplied-current Hall-MHD carrier trajectory."""
    if isinstance(n_steps, bool) or n_steps <= 0:
        raise ValueError("n_steps must be positive")
    states = [initial]
    state = initial
    for _ in range(n_steps):
        state = step_hall_mhd_pulsed(state, config, dt_s)
        states.append(state)
    return tuple(states)


def gkeyll_small_hall_acceptance_status() -> dict[str, str]:
    """Return the current fail-closed status for 2D external-code parity."""
    return {
        "case": "gkeyll_axisymmetric_small_hall",
        "status": "blocked_missing_public_same_case_reference",
        "required_artifact": "version-pinned Gkeyll/BOUT++ same-case output with provenance and checksum",
    }


def ono_fig4_acceptance_status() -> dict[str, str]:
    """Return the current fail-closed status for Ono figure reproduction."""
    return {
        "case": "ono_1997_fig4_flux_decay",
        "status": "blocked_missing_public_digitised_reference",
        "required_artifact": "digitised Ono flux-decay curve with deck metadata and checksum",
    }


def _implicit_be_step(
    psi: FloatArray,
    t_s: float,
    cfg: HallMHDPulsedConfig,
    dt_s: float,
) -> tuple[FloatArray, FloatArray]:
    rho = cfg.equilibrium.rho
    source = _ono_source(cfg, t_s + dt_s, rho)
    return np.asarray(
        (psi + dt_s * source) / (1.0 + dt_s / cfg.tau_psi_s), dtype=np.float64
    ), source


def _imex_rk2_step(
    psi: FloatArray,
    t_s: float,
    cfg: HallMHDPulsedConfig,
    dt_s: float,
) -> tuple[FloatArray, FloatArray]:
    rho = cfg.equilibrium.rho
    source_1 = _ono_source(cfg, t_s, rho)
    half_psi = (psi + 0.5 * dt_s * source_1) / (1.0 + 0.5 * dt_s / cfg.tau_psi_s)
    source_2 = _ono_source(cfg, t_s + 0.5 * dt_s, rho)
    del half_psi
    return np.asarray(
        (psi + dt_s * source_2) / (1.0 + dt_s / cfg.tau_psi_s), dtype=np.float64
    ), source_2


def _ono_source(cfg: HallMHDPulsedConfig, t_s: float, rho: FloatArray) -> FloatArray:
    e_theta = _resolve_e_theta(cfg, t_s)
    j_theta = _resolve_j_theta(cfg, t_s)
    return np.asarray(
        cfg.hall_scale * cfg.equilibrium.R_null * e_theta - _eta_profile(cfg, rho) * j_theta,
        dtype=np.float64,
    )


def _eta_profile(cfg: HallMHDPulsedConfig, rho: FloatArray) -> FloatArray:
    return spitzer_resistivity_ohm_m(
        np.full_like(rho, cfg.electron_temperature_eV),
        Z_eff=cfg.Z_eff,
        ln_lambda=cfg.ln_lambda,
    )


def _resolve_e_theta(cfg: HallMHDPulsedConfig, t_s: float) -> FloatArray:
    rho = cfg.equilibrium.rho
    if cfg.E_theta_t is None:
        return faraday_e_theta_from_b_ramp(rho, cfg.B_ext_t, t_s)
    return _coerce_profile("E_theta", cfg.E_theta_t(t_s, rho), rho.size)


def _resolve_j_theta(cfg: HallMHDPulsedConfig, t_s: float) -> FloatArray:
    rho = cfg.equilibrium.rho
    if cfg.J_theta_t is None:
        return np.asarray(cfg.equilibrium.J_theta, dtype=np.float64)
    return _coerce_profile("J_theta", cfg.J_theta_t(t_s, rho), rho.size)


def _magnetic_energy_proxy(rho: FloatArray, psi: FloatArray) -> float:
    return float(trapezoid(0.5 * psi * psi * 2.0 * np.pi * rho / MU_0, rho))


def _l2(rho: FloatArray, profile: FloatArray) -> float:
    return float(np.sqrt(max(trapezoid(profile * profile * 2.0 * np.pi * rho, rho), 0.0)))


def _validate_config(config: HallMHDPulsedConfig) -> HallMHDPulsedConfig:
    _validate_grid(config.equilibrium.rho)
    _validate_profile("equilibrium.psi", config.equilibrium.psi, config.equilibrium.rho.size)
    _require_positive("tau_psi_s", config.tau_psi_s)
    _require_positive("electron_temperature_eV", config.electron_temperature_eV)
    _require_positive("Z_eff", config.Z_eff)
    _require_positive("ln_lambda", config.ln_lambda)
    _require_finite("hall_scale", config.hall_scale)
    if config.closure not in {"axisymmetric_ono_flux", "adiabatic_electron"}:
        raise ValueError("closure must be 'axisymmetric_ono_flux' or 'adiabatic_electron'")
    if config.time_integrator not in {"implicit_be", "imex_rk2"}:
        raise ValueError("time_integrator must be 'implicit_be' or 'imex_rk2'")
    _require_finite("B_ext_t(0)", config.B_ext_t(0.0))
    return config


def _validate_state(state: HallMHDPulsedState, cfg: HallMHDPulsedConfig) -> None:
    _require_non_negative("state.t_s", state.t_s)
    _validate_profile("state.psi", state.psi, cfg.equilibrium.rho.size)
    _validate_profile("state.B_z", state.B_z, cfg.equilibrium.rho.size)


def _validate_grid(rho_m: FloatArray) -> FloatArray:
    rho = np.asarray(rho_m, dtype=np.float64)
    if rho.ndim != 1 or rho.size < 3:
        raise ValueError("rho_m must be a one-dimensional grid with at least three points")
    if np.any(~np.isfinite(rho)):
        raise ValueError("rho_m must be finite")
    if rho[0] < 0.0:
        raise ValueError("rho_m must start at a non-negative radius")
    if np.any(np.diff(rho) <= 0.0):
        raise ValueError("rho_m must be strictly increasing")
    return rho


def _validate_profile(name: str, values: FloatArray, expected_size: int) -> FloatArray:
    profile = np.asarray(values, dtype=np.float64)
    if profile.shape != (expected_size,):
        raise ValueError(f"{name} must have shape ({expected_size},)")
    if np.any(~np.isfinite(profile)):
        raise ValueError(f"{name} must be finite")
    return profile


def _coerce_profile(name: str, values: FloatArray | float, expected_size: int) -> FloatArray:
    profile = np.asarray(values, dtype=np.float64)
    if profile.shape == ():
        return np.full(expected_size, float(profile), dtype=np.float64)
    return _validate_profile(name, profile, expected_size)


def _require_positive(name: str, value: float) -> float:
    checked = _require_finite(name, value)
    if checked <= 0.0:
        raise ValueError(f"{name} must be positive")
    return checked


def _require_non_negative(name: str, value: float) -> float:
    checked = _require_finite(name, value)
    if checked < 0.0:
        raise ValueError(f"{name} must be non-negative")
    return checked


def _require_finite(name: str, value: float) -> float:
    checked = float(value)
    if not np.isfinite(checked):
        raise ValueError(f"{name} must be finite")
    return checked
